e2e/js_image/js_image_layer.bzl

```diff
@@ -0,0 +1,239 @@
+        destination = _runfile_path(ctx, file, runfiles_dir)
+        if _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            file_map[destination] = file
+
+    # executable and launcher should not go into runfiles directory so we add it to files here
+    files = depset([executable, launcher], transitive = [files])
+
+    symlinks = []
+
+    # NOTE: symlinks is different than root_symlinks. See: https://bazel.build/rules/rules#runfiles_symlinks for distinction between
+    # root_symlinks and symlinks and why they have to be handled differently.
+    for symlink in default.data_runfiles.symlinks.to_list():
+        destination = "/".join([runfiles_dir, ctx.workspace_name, symlink.path])
+        if not _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            continue
+        if hasattr(file_map, destination):
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    for symlink in default.data_runfiles.root_symlinks.to_list():
+        destination = "/".join([runfiles_dir, symlink.path])
+        if not _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            continue
+        if hasattr(file_map, destination):
+            file_map.pop(destination)
+        info = PackageSymlinkInfo(
+            target = "/%s" % _runfile_path(ctx, symlink.target_file, runfiles_dir),
+            destination = destination,
+            attributes = {"mode": "0777"},
+        )
+        symlinks.append([info, symlink.target_file.owner])
+
+    return [
+        PackageFilegroupInfo(
+            pkg_dirs = [],
+            pkg_files = [
+                [PackageFilesInfo(
+                    dest_src_map = file_map,
+                    attributes = {},
+                ), ctx.label],
+            ],
+            pkg_symlinks = symlinks,
+        ),
+        DefaultInfo(files = files),
+    ]
+
+runfiles = rule(
+    implementation = _runfiles_impl,
+    attrs = {
+        "binary": attr.label(mandatory = True),
+        "root": attr.string(),
+        "include": attr.string(),
+        "exclude": attr.string(),
+    },
+)
+
+BUILD_TAR = """
+import os
+import sys
+import json
+import subprocess
+from rules_python.python.runfiles import runfiles
+
+r = runfiles.Create()
+build_tar_path = r.Rlocation("rules_pkg/pkg/private/tar/build_tar")
+
+manifest_path = next((argv.replace("--manifest=", "") for argv in sys.argv if argv.startswith("--manifest=")), None)
+
+with open(manifest_path, 'r') as manifest_fp:
+    manifest = json.load(manifest_fp)
+
+def strip_execroot(p):
+    parts = p.split(os.sep)
+    i = parts.index("execroot") 
+    return os.sep.join(parts[i+2:]) 
+
+def get_runfiles_path(p):
+    for entry in manifest:
```

Comment:
This gets the `key` by `value` right? Shouldn't need to loop:

```
return list(manifest.keys())[list(manifest.values()).index(p)]
```

Make a list out of the keys in the dict, and index it by the list of values. `keys` and `values` are guaranteed to be in the same order (for the purposes here)

Reply:
manifest is a multidimensional array. I am not sure use of `.keys()` would work here.